Code/GraphMol/FileParsers/MaeMolSupplier.cpp

```diff
@@ -10,243 +10,424 @@
-    const auto ys = atom_data->getRealProperty("r_m_y_coord");
-    const auto zs = atom_data->getRealProperty("r_m_z_coord");
-    const auto atom_count = atomic_numbers->size();
-    shared_ptr<IndexedIntProperty> atomic_charges;
-    try {
-      atomic_charges = atom_data->getIntProperty("i_m_formal_charge");
-    } catch (std::out_of_range &) {
-    }
+void addAtoms(const mae::IndexedBlock &atom_block, RWMol &mol) {
+  // All atoms are guaranteed to have these three field names:
+  const auto atomic_numbers = atom_block.getIntProperty(mae::ATOM_ATOMIC_NUM);
+  const auto xs = atom_block.getRealProperty(mae::ATOM_X_COORD);
+  const auto ys = atom_block.getRealProperty(mae::ATOM_Y_COORD);
+  const auto zs = atom_block.getRealProperty(mae::ATOM_Z_COORD);
 
-    // atomic numbers, and x, y, and z coordinates
-    auto conf = new RDKit::Conformer(atom_count);
-    conf->set3D(true);
-    conf->setId(0);
-    PDBInfo pdb_info(atom_data);
-    for (size_t i = 0; i < atom_count; ++i) {
-      Atom *atom = new Atom(atomic_numbers->at(i));
-      mol.addAtom(atom, true, true);
-      if (atomic_charges) {
-        atom->setFormalCharge(atomic_charges->at(i));
-      }
-
-      RDGeom::Point3D pos;
-      pos.x = xs->at(i);
-      pos.y = ys->at(i);
-      pos.z = zs->at(i);
-      conf->setAtomPos(i, pos);
-
-      if(pdb_info.m_atom_name) {
-          addAtomPDBData(pdb_info, atom, i);
-      }
-    }
-    mol.addConformer(conf, false);
+  std::shared_ptr<mae::IndexedIntProperty> atomic_charges;
+
+  // atomic numbers, and x, y, and z coordinates
+  const auto size = atomic_numbers->size();
+  auto conf = new RDKit::Conformer(size);
+  conf->set3D(true);
+  conf->setId(0);
+
+  PDBInfo pdb_info(atom_block);
+
+  for (size_t i = 0; i < size; ++i) {
+    Atom *atom = new Atom(atomic_numbers->at(i));
+    mol.addAtom(atom, true, true);
+
+    pdb_info.addPDBData(atom, i);
+    set_atom_properties(*atom, atom_block, i);
+
+    RDGeom::Point3D pos;
+    pos.x = xs->at(i);
+    pos.y = ys->at(i);
+    pos.z = zs->at(i);
+    conf->setAtomPos(i, pos);
   }
+  mol.addConformer(conf, false);
+}
 
+void addBonds(const mae::IndexedBlock &bond_block, RWMol &mol) {
+  // All bonds are guaranteed to have these three field names:
+  const auto from_atoms = bond_block.getIntProperty(mae::BOND_ATOM_1);
+  const auto to_atoms = bond_block.getIntProperty(mae::BOND_ATOM_2);
+  const auto orders = bond_block.getIntProperty(mae::BOND_ORDER);
+  const auto size = from_atoms->size();
+
+  for (size_t i = 0; i < size; ++i) {
+    // Maestro atoms are 1 indexed!
+    const auto from_atom = from_atoms->at(i) - 1;
+    const auto to_atom = to_atoms->at(i) - 1;
+    const auto order = bolookup.find(orders->at(i))->second;
+    if (from_atom > to_atom) continue;  // Maestro files double-list bonds
+
+    auto bond = new Bond(order);
+    bond->setOwningMol(mol);
+    bond->setBeginAtomIdx(from_atom);
+    bond->setEndAtomIdx(to_atom);
+    mol.addBond(bond, true);
```

Comment:
I've only used:

    mol.addBond(from_atom, to_atom, order);

Is there a reason to prefer one idiom over the other?

Reply:
`mol.addBond(from_atom, to_atom, order);` can't take indexes for `from_atom` and `to_atom`, they have to be `Atom*`.